### PR TITLE
[sparsity] WeightNormSparsifier

### PR DESCRIPTION
**Pull request by z-a-f**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #60728 [quant][sparsity] Generic convert function
* #59771 [sparsity] Lambda Scheduler
* #59770 [sparsity] Base sparsity level scheduler class
* **#58955 [sparsity] WeightNormSparsifier**
* #58707 [sparsity][refactor] Import factoring out
* #58704 [sparsity] Sparsifier class
* #58705 [sparsity] Sparsity parametrization
* #60850 [sparsity][refactor] Changing linear row/col control
* #60887 [sparsity] Add sparsity tests to run_test.py

Implements the weight norm sparsifier.
This type of sparsifier computes the norm of the weights, sorts them, and zeroes-out the target fraction of them.

The main imeplemented method is `update_mask`, which holds the main logic of changing the masks.

Test Plan:

```
python test/test_ao_sparsity.py
```

Differential Revision: [D28970960](https://our.internmc.facebook.com/intern/diff/D28970960)

### CODE REVIEW

File: torch/ao/sparsity/sparsifier/weight_norm_sparsifier.py [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+    def __init__(self, config, sparsity_level=0.5, sparsity_pattern=(1, 4)):
+        defaults = {
+            'sparsity_level': sparsity_level,
+            'sparsity_pattern': sparsity_pattern
+        }
+        super().__init__(config, defaults)
+
+    def update_mask(self, layer, sparsity_level, sparsity_pattern, **kwargs):
+        if sparsity_level <= 0:
+            layer.mask = torch.ones(layer.weight.shape)
```
Comment (vadimkantorov) [inline torch/ao/sparsity/sparsifier/weight_norm_sparsifier.py:18]: layer.mask could poptentially be on another device than `layer.weight`, right?

**Comment by z-a-f** [comment]
@zafartahirov has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28970960).

**Comment by z-a-f** [comment]
@zafartahirov has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28970960).

**Comment by z-a-f** [comment]
@zafartahirov has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28970960).

**Comment by z-a-f** [comment]
@zafartahirov has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28970960).

File: torch/ao/sparsity/sparsifier/weight_norm_sparsifier.py [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+    def update_mask(self, layer, sparsity_level, sparsity_pattern,
+                    zeros_per_block, **kwargs):
+        if zeros_per_block != reduce((lambda x, y: x * y), sparsity_pattern):
+            raise NotImplementedError('Partial block sparsity is not yet there')
+        if sparsity_level <= 0:
+            layer.mask.data = torch.ones(layer.weight.shape, device=layer.weight.device)
+        elif sparsity_level >= 1.0:
+            layer.mask.data = torch.zeros(layer.weight.shape, device=layer.weight.device)
+        else:
+            ww = layer.weight * layer.weight
```
Comment (z-a-f) [inline torch/ao/sparsity/sparsifier/weight_norm_sparsifier.py:35]:
```
original_weight = layer.parametrizations.weight[0].original
ww = original_weight * original_weight
```

**Comment by z-a-f** [comment]
@zafartahirov has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28970960).

File: torch/ao/sparsity/sparsifier/weight_norm_sparsifier.py [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+    rows = idx // shape[1]
+    cols = idx % shape[1]
+    return rows, cols
+
+class WeightNormSparsifier(BaseSparsifier):
+    def __init__(self,
+                 sparsity_level=0.5, sparsity_pattern=(1, 4),
+                 zeros_per_block=None):
+        if zeros_per_block is None:
+            zeros_per_block = reduce((lambda x, y: x * y), sparsity_pattern)
```
Comment (raghuramank100) [inline torch/ao/sparsity/sparsifier/weight_norm_sparsifier.py:18]: Is this restricted to 2D weights?

File: torch/ao/sparsity/sparsifier/base_sparsifier.py [hunk 4]
```diff
@@ -177,6 +178,14 @@ def convert(self):
...
+            return
+        with torch.no_grad():
+            for config in self.module_groups:
+                module = config['module']
+                self.update_mask(module, **config)
+
     @abc.abstractmethod
-    def step(self):
-        return
+    def update_mask(self, layer, **kwargs):
```
Comment (raghuramank100) [inline torch/ao/sparsity/sparsifier/base_sparsifier.py:190]: Lets call this function as step instead of update mask. The reason is that:
 For every call of forward, we want the sparsifier to be called. This is to allow for updating of any stats used for mask calculation: This could be gradients, weight norms, activations etc.
The rate at  which the mask is updated can be controlled by the scheduler. At a high level, we do this:
``` python
def step(self):
  # Always update stats
  # Depending on scheduler, call self.update_mask
```

File: torch/ao/sparsity/sparsifier/weight_norm_sparsifier.py [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+        super().__init__(defaults=defaults)
+
+    def update_mask(self, layer, sparsity_level, sparsity_pattern,
+                    zeros_per_block, **kwargs):
+        if zeros_per_block != reduce((lambda x, y: x * y), sparsity_pattern):
+            raise NotImplementedError('Partial block sparsity is not yet there')
+        # TODO: Add support for multiple parametrizations for the same weight
+        mask = layer.parametrizations.weight[0].mask
+        if sparsity_level <= 0:
+            mask.data = torch.ones(layer.weight.shape, device=layer.weight.device)
```
Comment (raghuramank100) [inline torch/ao/sparsity/sparsifier/weight_norm_sparsifier.py:33]: Why mask.data instead of mask here?

File: torch/ao/sparsity/sparsifier/weight_norm_sparsifier.py [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+            sorted_idx = sorted_idx[:threshold_idx]
+            rows, cols = _flat_idx_to_2d(sorted_idx, ww_pool.shape[1:])
+            rows *= sparsity_pattern[0]
+            cols *= sparsity_pattern[1]
+
+            new_mask = torch.ones(ww.shape, device=layer.weight.device)
+            for row, col in zip(rows, cols):
+                new_mask[row:row + sparsity_pattern[0],
+                         col:col + sparsity_pattern[1]] = 0
+            mask.data *= new_mask
```
Comment (raghuramank100) [inline torch/ao/sparsity/sparsifier/weight_norm_sparsifier.py:53]: Why are we updating mask.data here? Why not mask=new_mask?

File: torch/ao/sparsity/sparsifier/base_sparsifier.py [hunk 7]
```diff
@@ -177,6 +178,14 @@ def convert(self):
         raise NotImplementedError('`convert` is not implemented. '
             'Please, use `torch.ao.utils.convert` instead.')
 
+    def step(self, use_path=True):
+        if not self.enable_mask_update:
+            return
+        with torch.no_grad():
+            for config in self.module_groups:
+                module = config['module']
+                self.update_mask(module, **config)
```
Comment (raghuramank100) [inline torch/ao/sparsity/sparsifier/base_sparsifier.py:187]: NIT: Since update_mask is a member of the sparsifier object, why have it take arguments?

File: test/ao/sparsity/test_sparsifier.py [hunk 8]
```diff
@@ -45,6 +52,14 @@ def test_constructor(self):
         assert 'test' in sparsifier.module_groups[0]
         assert sparsifier.module_groups[0]['test'] == 3
 
+    def test_step(self):
+        model = Model()
+        sparsifier = ImplementedSparsifier(test=3)
+        sparsifier.enable_mask_update = True
```
Comment (raghuramank100) [inline test/ao/sparsity/test_sparsifier.py:58]: By default, can enable_mask_update be True, so that in the absence of a scheduler, the mask is updated with every call to step? This way we can avoid requiring users to set this attribute as an additional step.

File: test/ao/sparsity/test_sparsifier.py [hunk 9]
```diff
@@ -55,3 +70,37 @@ def test_state_dict(self):
         assert sparsifier0.module_groups != sparsifier1.module_groups
         sparsifier1.load_state_dict(state_dict)
         assert sparsifier0.module_groups == sparsifier1.module_groups
+
+    def test_mask_squash(self):
+        model = Model()
+        sparsifier = ImplementedSparsifier(test=3)
+        sparsifier.prepare(model, [model.linear])
+        assert hasattr(model.linear, 'mask')
```
Comment (raghuramank100) [inline test/ao/sparsity/test_sparsifier.py:78]: This shouldnt be the case isn't it? Isnt mask an attr of the fake-sparse module?

File: test/ao/sparsity/test_sparsifier.py [hunk 10]
```diff
@@ -55,3 +70,37 @@ def test_state_dict(self):
...
+    def test_constructor(self):
+        model = Model()
+        sparsifier = WeightNormSparsifier()
+        sparsifier.prepare(model, config=None)
+        for g in sparsifier.module_groups:
+            assert isinstance(g['module'], nn.Linear)
+            # The module_groups are unordered
+            assert g['path'] in ('seq.0', 'linear')
+
+    def test_logic(self):
```
Comment (raghuramank100) [inline test/ao/sparsity/test_sparsifier.py:100]: Can we call this test_step()?

File: test/ao/sparsity/test_sparsifier.py [hunk 11]
```diff
@@ -55,3 +70,37 @@ def test_state_dict(self):
...
+        model = Model()
+        sparsifier = WeightNormSparsifier(sparsity_level=0.5)
+        sparsifier.prepare(model, config=[model.linear])
+        sparsifier.enable_mask_update = True
+        sparsifier.step()
+        self.assertAlmostEqual(model.linear.mask.mean().item(), 0.5, places=2)
```
Comment (raghuramank100) [inline test/ao/sparsity/test_sparsifier.py:106]: Do we need a test to check that the masks are serialized correctly as part of the model?

**Comment by z-a-f** [comment]
@zafartahirov has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28970960).

**Comment by z-a-f** [comment]
@zafartahirov has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28970960).

File: test/ao/sparsity/test_sparsifier.py (hunk 8 above)
Comment (z-a-f) [reply test/ao/sparsity/test_sparsifier.py:58]: Part of the base sparsifier PR

File: test/ao/sparsity/test_sparsifier.py (hunk 11 above)
Comment (z-a-f) [reply test/ao/sparsity/test_sparsifier.py:106]: I thought those are tested when I test the parametrization state tests.

File: torch/ao/sparsity/sparsifier/base_sparsifier.py (hunk 7 above)
Comment (z-a-f) [reply torch/ao/sparsity/sparsifier/base_sparsifier.py:187]: Oh that's because the `update_mask` is just a logic of changing generic masks. The config specifies which layers is the update called for. I am doing it so that the user would need to write minimum amount of code.

File: torch/ao/sparsity/sparsifier/weight_norm_sparsifier.py (hunk 3 above)
Comment (z-a-f) [reply torch/ao/sparsity/sparsifier/weight_norm_sparsifier.py:18]: No, the reduce can take care of any dimensions.

File: torch/ao/sparsity/sparsifier/weight_norm_sparsifier.py (hunk 5 above)
Comment (z-a-f) [reply torch/ao/sparsity/sparsifier/weight_norm_sparsifier.py:33]: This is a constraint that we must have, otherwise a new variable will be created

File: torch/ao/sparsity/sparsifier/weight_norm_sparsifier.py (hunk 6 above)
Comment (z-a-f) [reply torch/ao/sparsity/sparsifier/weight_norm_sparsifier.py:53]: We are not allowed to use non-inplace assignments. All assignments to the mask must be in-place

**Comment by z-a-f** [comment]
@zafartahirov has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28970960).

**Comment by z-a-f** [comment]
@zafartahirov has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28970960).

**Comment by z-a-f** [comment]
@zafartahirov has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28970960).

**Review by raghuramank100 — APPROVED** [review]

**Comment by z-a-f** [comment]
@zafartahirov has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28970960).

**Comment by z-a-f** [comment]
@zafartahirov has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28970960).
